lms/static/scripts/frontend_apps/components/test/BookSelector-test.js

```diff
@@ -0,0 +1,224 @@
+    input.getDOMNode().value = url;
+  };
+
+  // Set the value of the input field AND fire `change` event
+  const updateURL = (wrapper, url) => {
+    setURL(wrapper, url);
+    const input = getInput(wrapper);
+    input.simulate('change');
+  };
+
+  context('entering, changing and submitting book URL', () => {
+    it('validates entered URL when the value of the text input changes', () => {
+      const wrapper = renderBookSelector();
+
+      updateURL(wrapper, 'foo');
+
+      assert.calledOnce(fakeBookIDFromURL);
+      assert.calledWith(fakeBookIDFromURL, 'foo');
+
+      updateURL(wrapper, 'bar');
+
+      assert.equal(
+        fakeBookIDFromURL.callCount,
+        2,
+        're-validates if entered URL value changes'
+      );
+      assert.calledWith(fakeBookIDFromURL, 'bar');
+
+      updateURL(wrapper, 'bar');
+
+      assert.equal(
+        fakeBookIDFromURL.callCount,
+        2,
+        'Does not validate URL if it has not changed from previous value'
+      );
+    });
+
+    it('validates entered URL when input is focused and "Enter" is pressed', () => {
+      const wrapper = renderBookSelector();
+      const input = getInput(wrapper);
+      const keyEvent = new Event('keydown');
+      keyEvent.key = 'Enter';
+
+      setURL(wrapper, 'http://www.example.com');
+
+      input.getDOMNode().dispatchEvent(keyEvent);
+
+      assert.calledOnce(fakeBookIDFromURL);
+      assert.calledWith(fakeBookIDFromURL, 'http://www.example.com');
+    });
+
+    it('validates entered URL when `IconButton` is clicked', () => {
+      const wrapper = renderBookSelector();
+      setURL(wrapper, 'foo');
+
+      wrapper.find('IconButton button[title="Find book"]').simulate('click');
+
+      assert.calledOnce(fakeBookIDFromURL);
+      assert.calledWith(fakeBookIDFromURL, 'foo');
+    });
+
+    it('does not attempt to check the URL format if the field value is empty', () => {
+      const wrapper = renderBookSelector();
+      updateURL(wrapper, 'foo');
+
+      assert.calledOnce(fakeBookIDFromURL);
+
+      updateURL(wrapper, '');
+
+      assert.calledOnce(fakeBookIDFromURL);
+    });
+
+    it('shows an error if entered URL format is invalid', () => {
+      fakeBookIDFromURL.returns(null);
+
+      const wrapper = renderBookSelector();
+      updateURL(wrapper, 'foo');
+
+      const errorMessage = wrapper.find('[data-testid="error-message"]');
+
+      assert.isTrue(errorMessage.exists());
+      assert.include(
+        errorMessage.text(),
+        "That doesn't look like a VitalSource book link"
+      );
+      assert.isTrue(errorMessage.find('SvgIcon[name="cancel"]').exists());
+    });
+  });
+
+  context('correctly-formatted URL entered', () => {
+    it('fetches book metadata associated with the entered URL', async () => {
+      const onSelectBook = sinon.stub();
+      const wrapper = renderBookSelector({ onSelectBook });
+      updateURL(wrapper, 'a valid URL');
+
+      await waitFor(() => onSelectBook.called);
```

Comment:
It would be really useful if Sinon stubs had a property like `called` that returned a promise for when the stub is called.

Reply:
Wouldn't it?! Also, the `sinon` chai assertions don't take an optional `message` argument, something I learned yesterday. Lots of fiddly bits to these particular tests; I hope that doesn't indicate an underlying smelliness.